File: examples/python/documentation/iterators.py
# Copyright (c) HashiCorp, Inc.
# SPDX-License-Identifier: MPL-2.0

from cdktf import TerraformStack, Token, TerraformCount, TerraformAsset
from constructs import Construct
from imports.aws.provider import AwsProvider
from imports.aws.instance import Instance
from imports.github.data_github_organization import DataGithubOrganization
from imports.github.provider import GithubProvider
from imports.github.team import Team
from imports.github.team_members import TeamMembers
from imports.aws.s3_bucket_object import S3BucketObject
from imports.aws.acm_certificate import AcmCertificate
from imports.aws.acm_certificate_validation import AcmCertificateValidation
from imports.aws.data_aws_route53_zone import DataAwsRoute53Zone
from imports.aws.route53_record import Route53Record


# DOCS_BLOCK_START:iterators-define-iterators,iterators-iterators-complex-types
from imports.aws.s3_bucket import S3Bucket
from cdktf import TerraformIterator, TerraformVariable, TerraformLocal
# DOCS_BLOCK_END:iterators-define-iterators,iterators-iterators-complex-types


class IteratorStackOne(TerraformStack):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)
        AwsProvider(self, "aws", region="us-east-1")

        # DOCS_BLOCK_START:iterators-iterators-complex-types
        map = TerraformLocal(self, "my-map", {
            "website": {
                "name": "website-static-files",
                "tags": {"app": "website"}
            },
            "images": {
                "name": "images",
                "tags": {"app": "image-converter"}
            }
        })

        iterator = TerraformIterator.from_map(
            map=map.as_any_map
        )

        s3Bucket = S3Bucket(self, "s3-bucket",
                            for_each=iterator,
                            bucket=iterator.get_string("name"),
                            tags=iterator.get_map("tags")
                            )
        # DOCS_BLOCK_END:iterators-iterators-complex-types

        # DOCS_BLOCK_START:iterators-count
        servers = TerraformVariable(self, "servers",
                                    type="number"
                                    )

        count = TerraformCount.of(servers.number_value)

        Instance(self, "server",
                 count=count,
                 ami="ami-a1b2c3d4",
                 instance_type="t2.micro",
                 tags={
                     "Name": "Server ${" + Token().as_string(count.index) + "}"
                 }
                 )
        # DOCS_BLOCK_END:iterators-count


class IteratorStackTwo(TerraformStack):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)
        AwsProvider(self, "aws", region="us-east-1")
        GithubProvider(self, "gh")
        # DOCS_BLOCK_START:iterators-define-iterators

        list = TerraformVariable(self, "list",
                                 type="list(string)"
                                 )

        iterator = TerraformIterator.from_list(list=list.list_value)

        s3Bucket = S3Bucket(self, "bucket",
                            for_each=iterator,
                            bucket=Token.as_string(iterator.value)
                            )
        # DOCS_BLOCK_END:iterators-define-iterators

        # DOCS_BLOCK_START:iterators-list-attributes
        org_name = "my-org"

        GithubProvider(self, "github",
                       organization=org_name
                       )

        github_team = Team(self, "core-team",
                           name="core"
                           )

        org_members = DataGithubOrganization(self, "org",
                                             name=org_name
                                             )

        org_member_iterator = TerraformIterator.from_list(org_members.members)

        TeamMembers(self, "members",
                    team_id=github_team.id,
                    members=org_member_iterator.dynamic({
                        "username": Token().as_string(org_member_iterator.value),
                        "role": "maintainer"
                    })
                    )
        # DOCS_BLOCK_END:iterators-list-attributes
        # DOCS_BLOCK_START:iterators-complex-lists
        cert = AcmCertificate(self, "cert",
                              domain_name="example.com",
                              validation_method="DNS",
                              )
        data_aws_route53_zone_example = DataAwsRoute53Zone(self, "dns_zone",
                                                           name="example.com",
                                                           private_zone=False
                                                           )

        example_for_each_iterator = TerraformIterator.from_complex_list(
            cert.domain_validation_options,
            "domain_name"
        )

        records = Route53Record(self, "record",
                                for_each=example_for_each_iterator,
                                allow_overwrite=True,
                                name=example_for_each_iterator.get_string(
                                    "resource_record_name"),
                                records=[
                                    example_for_each_iterator.get_string("resource_record_record")],
                                ttl=60,
                                type=example_for_each_iterator.get_string(
                                    "resource_record_type"),
                                zone_id=data_aws_route53_zone_example.zone_id,
                                )

        records_iterator = TerraformIterator.from_resources(records)

        AcmCertificateValidation(self, "validation",
                                 certificate_arn=cert.arn,
                                 validation_record_fqdns=Token.as_list(
                                     records_iterator.pluck_property(
                                         "fqdn")
                                 ),
                                 )
        # DOCS_BLOCK_END:iterators-complex-lists

        # DOCS_BLOCK_START:iterators-chain
        map = TerraformLocal(self, "my-map", {
            "website": {
                "name": "website-static-files",
                "tags": {"app": "website"}
            },
            "images": {
                "name": "images",
                "tags": {"app": "image-converter"}
            }
        })
        s3_bucket_configuration_iterator = TerraformIterator.from_map(
            map=map.as_any_map
        )
        s3_buckets = S3Bucket(self, "complex-iterator-buckets",
                              for_each=s3_bucket_configuration_iterator,
                              bucket=s3_bucket_configuration_iterator.get_string(
                                  "name"),
                              tags=s3_bucket_configuration_iterator.get_map(
                                  "tags")
                              )

        # This would be TerraformIterator.from_data_sources for data_sources
        s3_buckets_iterator = TerraformIterator.from_resources(s3_buckets)
        help_file = TerraformAsset(self, "help",
                                   path="./help"
                                   )
        S3BucketObject(self, "object",
                       for_each=s3_buckets_iterator,
                       bucket=s3_buckets_iterator.get_string("id"),
                       key="help",
                       source=help_file.path
                       )
        # DOCS_BLOCK_END:iterators-chain

        # DOCS_BLOCK_START:iterators-for-expression
        values = TerraformLocal(self, "map-local", {
            "website": {
                "name": "website-static-files",
                "tags": {"app": "website"}
            },
            "images": {
                "name": "images",
                "tags": {"app": "image-converter"}
            }
        })
        mapIterator = TerraformIterator.from_map(
            map=values.as_any_map
        )
        TerraformLocal(self, "list-of-keys", mapIterator.keys())
        TerraformLocal(self, "list-of-values", mapIterator.values())
        TerraformLocal(self, "list-of-names",
                       mapIterator.pluck_property("name"))
        TerraformLocal(self, "list-of-names-of-included",
                       mapIterator.for_expression_for_list("val.name if val.included"))
        TerraformLocal(self, "map-with-names-as-key-and-tags-as-value-of-included",
                       mapIterator.for_expression_for_map("val.name", "val.tags if val.included"))
        # DOCS_BLOCK_END:iterators-for-expression

File: examples/typescript/documentation/iterators.ts
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { GithubProvider } from "@cdktf/provider-github/lib/provider";
import { Team } from "@cdktf/provider-github/lib/team";
import { DataGithubOrganization } from "@cdktf/provider-github/lib/data-github-organization";
import { TeamMembers } from "@cdktf/provider-github/lib/team-members";
// DOCS_BLOCK_START:iterators,iterators-complex-types
import {
  TerraformIterator,
  TerraformLocal,
  TerraformOutput,
  TerraformStack,
  TerraformVariable,
} from "cdktf";
import { Construct } from "constructs";
import { AwsProvider } from "@cdktf/provider-aws/lib/aws-provider";
import { S3Bucket } from "@cdktf/provider-aws/lib/s3-bucket";

// DOCS_BLOCK_END:iterators,iterators-complex-types
import { TerraformAsset } from "cdktf";
import { S3BucketObject } from "@cdktf/provider-aws/lib/s3-bucket-object";
// DOCS_BLOCK_START:iterators,iterators-complex-types
export class IteratorsStack extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new AwsProvider(this, "aws", {
      region: "us-west-2",
    });

    // DOCS_BLOCK_END:iterators,iterators-complex-types

    // DOCS_BLOCK_START:iterators-list-attributes
    const orgName = "my-org";

    new GithubProvider(this, "github", {
      organization: orgName,
    });

    const team = new Team(this, "core-team", {
      name: "core",
    });

    const orgMembers = new DataGithubOrganization(this, "org", {
      name: orgName,
    });

    const orgMemberIterator = TerraformIterator.fromList(orgMembers.members);

    new TeamMembers(this, "members", {
      teamId: team.id,
      members: orgMemberIterator.dynamic({
        username: orgMemberIterator.value,
        role: "maintainer",
      }),
    });
    // DOCS_BLOCK_END:iterators-list-attributes

    // DOCS_BLOCK_START:iterators
    const list = new TerraformVariable(this, "list", {
      type: "list(string)",
    });

    const simpleIterator = TerraformIterator.fromList(list.listValue);

    new S3Bucket(this, "iterator-bucket", {
      forEach: simpleIterator,
      bucket: simpleIterator.value,
    });
    // DOCS_BLOCK_END:iterators

    // DOCS_BLOCK_START:iterators-complex-types
    const complexIterator = TerraformIterator.fromMap({
      website: {
        name: "website-static-files",
        tags: { app: "website" },
      },
      images: {
        name: "images",
        tags: { app: "image-converter" },
      },
    });

    new S3Bucket(this, "complex-iterator-bucket", {
      forEach: complexIterator,
      bucket: complexIterator.getString("name"),
      tags: complexIterator.getStringMap("tags"),
    });
    // DOCS_BLOCK_END:iterators-complex-types

    // DOCS_BLOCK_START:iterators-chain
    const s3BucketConfigurationIterator = TerraformIterator.fromMap({
      website: {
        name: "website-static-files",
        tags: { app: "website" },
      },
      images: {
        name: "images",
        tags: { app: "image-converter" },
      },
    });

    const s3Buckets = new S3Bucket(this, "complex-iterator-buckets", {
      forEach: s3BucketConfigurationIterator,
      bucket: s3BucketConfigurationIterator.getString("name"),
      tags: s3BucketConfigurationIterator.getStringMap("tags"),
    });

    // This would be TerraformIterator.fromDataSources for data_sources
    const s3BucketsIterator = TerraformIterator.fromResources(s3Buckets);
    const helpFile = new TerraformAsset(this, "help", {
      path: "./help",
    });
    new S3BucketObject(this, "object", {
      forEach: s3BucketsIterator,
      bucket: s3BucketsIterator.getString("id"),
      key: "help",
      source: helpFile.path,
    });
    // DOCS_BLOCK_END:iterators-chain

    // DOCS_BLOCK_START:iterators-for-expression
    const mapIterator = TerraformIterator.fromMap({
      website: {
        name: "website-static-files",
        tags: { app: "website" },
        included: true,
      },
      images: {
        name: "images",
        tags: { app: "image-converter" },
      },
    });
    new TerraformLocal(this, "list-of-keys", mapIterator.keys());
    new TerraformLocal(this, "list-of-values", mapIterator.values());
    new TerraformLocal(
      this,
      "list-of-names",
      mapIterator.pluckProperty("name")
    );
    new TerraformLocal(
      this,
      "list-of-names-of-included",
      mapIterator.forExpressionForList("val.name if val.included")
    );
    new TerraformLocal(
      this,
      "map-with-names-as-key-and-tags-as-value-of-included",
      mapIterator.forExpressionForMap("val.name", "val.tags if val.included")
    );
    // DOCS_BLOCK_END:iterators-for-expression

    // DOCS_BLOCK_START:iterators,iterators-complex-types
  }
}
// DOCS_BLOCK_END:iterators,iterators-complex-types
